File: ledger-models-java/src/test/java/common/models/security/bonds/MortgageBackedSecurityFromPricerInputsTest.java
package common.models.security.bonds;

import common.models.security.CashSecurity;
import fintekkers.models.security.CouponFrequencyProto;
import fintekkers.models.security.CouponTypeProto;
import fintekkers.models.security.ProductTypeProto;
import fintekkers.models.security.SecurityProto;
import fintekkers.models.security.bond.AgencyProto;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.ZonedDateTime;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class MortgageBackedSecurityFromPricerInputsTest {

    private static final BigDecimal FACE = new BigDecimal("1000");
    private static final BigDecimal COUPON = new BigDecimal("0.04");
    private static final LocalDate ISSUE = LocalDate.of(2024, 2, 1);
    private static final LocalDate MATURITY = LocalDate.of(2054, 1, 1);

    private static final String POOL_NUMBER = "FN AS1234";
    private static final AgencyProto AGENCY = AgencyProto.FNMA;
    private static final BigDecimal WAC = new BigDecimal("0.045");
    private static final int WAM = 358;
    private static final BigDecimal PASS_THROUGH_RATE = new BigDecimal("0.04");
    private static final BigDecimal CURRENT_FACTOR = new BigDecimal("0.95");
    private static final BigDecimal ORIGINAL_FACE_VALUE = new BigDecimal("250000000");
    private static final BigDecimal CURRENT_UPB = new BigDecimal("237500000");
    private static final BigDecimal PSA_SPEED = new BigDecimal("150");

    @Test
    public void mbsFromPricerInputs_populatesBondDetailsAndMbsExtension() {
        SecurityProto proto = MortgageBackedSecurity.fromPricerInputs(
                FACE, COUPON, CouponTypeProto.FIXED, CouponFrequencyProto.MONTHLY,
                ISSUE, MATURITY,
                POOL_NUMBER, AGENCY, WAC, WAM, PASS_THROUGH_RATE,
                CURRENT_FACTOR, ORIGINAL_FACE_VALUE, CURRENT_UPB, PSA_SPEED);

        assertEquals(ProductTypeProto.MORTGAGE_BACKED, proto.getProductType());
        assertTrue(proto.hasBondDetails(), "bond_details must be populated");
        assertTrue(proto.hasMbsExtension(), "mbs_extension must be populated");

        assertEquals(FACE.toPlainString(),
                proto.getBondDetails().getFaceValue().getArbitraryPrecisionValue());
        assertEquals(COUPON.toPlainString(),
                proto.getBondDetails().getCouponRate().getArbitraryPrecisionValue());

        assertEquals(POOL_NUMBER, proto.getMbsExtension().getPoolNumber());
        assertEquals(AGENCY, proto.getMbsExtension().getAgency());
        assertEquals(WAC.toPlainString(),
                proto.getMbsExtension().getWac().getArbitraryPrecisionValue());
        assertEquals(WAM, proto.getMbsExtension().getWam());
        assertEquals(PASS_THROUGH_RATE.toPlainString(),
                proto.getMbsExtension().getPassThroughRate().getArbitraryPrecisionValue());
        assertEquals(CURRENT_FACTOR.toPlainString(),
                proto.getMbsExtension().getCurrentFactor().getArbitraryPrecisionValue());
        assertEquals(ORIGINAL_FACE_VALUE.toPlainString(),
                proto.getMbsExtension().getOriginalFaceValue().getArbitraryPrecisionValue());
        assertEquals(CURRENT_UPB.toPlainString(),
                proto.getMbsExtension().getCurrentUpb().getArbitraryPrecisionValue());
        assertEquals(PSA_SPEED.toPlainString(),
                proto.getMbsExtension().getPsaSpeed().getArbitraryPrecisionValue());
    }

    @Test
    public void mbsFromPricerInputs_roundTripsViaBytes() throws Exception {
        SecurityProto original = MortgageBackedSecurity.fromPricerInputs(
                FACE, COUPON, CouponTypeProto.FIXED, CouponFrequencyProto.MONTHLY,
                ISSUE, MATURITY,
                POOL_NUMBER, AGENCY, WAC, WAM, PASS_THROUGH_RATE,
                CURRENT_FACTOR, ORIGINAL_FACE_VALUE, CURRENT_UPB, PSA_SPEED);

        byte[] bytes = original.toByteArray();
        SecurityProto decoded = SecurityProto.parseFrom(bytes);

        assertEquals(original, decoded);
        assertEquals(ProductTypeProto.MORTGAGE_BACKED, decoded.getProductType());

        // Re-wrap into the typed domain object and verify each typed accessor.
        MortgageBackedSecurity wrapper = new MortgageBackedSecurity(
                UUID.randomUUID(), "Test", ZonedDateTime.now(), CashSecurity.USD);
        wrapper.setSecurityProto(decoded);

        assertEquals(POOL_NUMBER, wrapper.getPoolNumber());
        assertEquals(AGENCY, wrapper.getAgency());
        assertEquals(0, WAC.compareTo(wrapper.getWac()));
        assertEquals(WAM, wrapper.getWam());
        assertEquals(0, PASS_THROUGH_RATE.compareTo(wrapper.getPassThroughRate()));
        assertEquals(0, CURRENT_FACTOR.compareTo(wrapper.getCurrentFactor()));
        assertEquals(0, ORIGINAL_FACE_VALUE.compareTo(wrapper.getOriginalFaceValue()));
        assertEquals(0, CURRENT_UPB.compareTo(wrapper.getCurrentUpb()));
        assertEquals(0, PSA_SPEED.compareTo(wrapper.getPsaSpeed()));
    }
}

File: ledger-models-javascript/node/wrappers/models/security/MortgageBackedSecurity.fromPricerInputs.test.ts
import Security from './security';
import BondSecurity from './BondSecurity';
import MortgageBackedSecurity from './MortgageBackedSecurity';
import { SecurityProto } from '../../../fintekkers/models/security/security_pb';
import { ProductTypeProto } from '../../../fintekkers/models/security/product_type_pb';
import { CouponTypeProto } from '../../../fintekkers/models/security/coupon_type_pb';
import { CouponFrequencyProto } from '../../../fintekkers/models/security/coupon_frequency_pb';
import { AgencyProto } from '../../../fintekkers/models/security/bond/agency_pb';
import { LocalDate } from '../utils/date';
import { LocalDateProto } from '../../../fintekkers/models/util/local_date_pb';
import { Decimal } from 'decimal.js';

function makeDate(y: number, m: number, d: number): LocalDate {
  return new LocalDate(new LocalDateProto().setYear(y).setMonth(m).setDay(d));
}

const baseInputs = {
  faceValue: new Decimal('250000000'),
  couponRate: new Decimal('0.04'),
  couponType: CouponTypeProto.FIXED,
  couponFrequency: CouponFrequencyProto.MONTHLY,
  issueDate: makeDate(2024, 1, 1),
  maturityDate: makeDate(2054, 1, 1),
};

const mbsInputs = {
  ...baseInputs,
  poolNumber: 'FN AS1234',
  agency: AgencyProto.FNMA,
  wac: new Decimal('0.045'),
  wam: 358,
  passThroughRate: new Decimal('0.04'),
  currentFactor: new Decimal('0.95'),
  originalFaceValue: new Decimal('250000000'),
  currentUpb: new Decimal('237500000'),
  psaSpeed: new Decimal('150'),
};

test('MortgageBackedSecurity.fromPricerInputs populates bond_details and mbs_extension', () => {
  const proto = MortgageBackedSecurity.fromPricerInputs(mbsInputs);
  expect(proto.getProductType()).toBe(ProductTypeProto.MORTGAGE_BACKED);
  expect(proto.hasBondDetails()).toBe(true);
  expect(proto.hasMbsExtension()).toBe(true);

  const ext = proto.getMbsExtension()!;
  expect(ext.getPoolNumber()).toBe('FN AS1234');
  expect(ext.getAgency()).toBe(AgencyProto.FNMA);
  expect(ext.getWac()?.getArbitraryPrecisionValue()).toBe('0.045');
  expect(ext.getWam()).toBe(358);
  expect(ext.getPassThroughRate()?.getArbitraryPrecisionValue()).toBe('0.04');
  expect(ext.getCurrentFactor()?.getArbitraryPrecisionValue()).toBe('0.95');
  expect(ext.getOriginalFaceValue()?.getArbitraryPrecisionValue()).toBe('250000000');
  expect(ext.getCurrentUpb()?.getArbitraryPrecisionValue()).toBe('237500000');
  expect(ext.getPsaSpeed()?.getArbitraryPrecisionValue()).toBe('150');
});

test('MortgageBackedSecurity round-trips via serializeBinary / deserializeBinary preserving all 9 mbs fields', () => {
  const proto = MortgageBackedSecurity.fromPricerInputs(mbsInputs);
  const bytes = proto.serializeBinary();
  const round = SecurityProto.deserializeBinary(bytes);

  expect(round.getProductType()).toBe(ProductTypeProto.MORTGAGE_BACKED);
  expect(round.hasBondDetails()).toBe(true);
  expect(round.hasMbsExtension()).toBe(true);

  const ext = round.getMbsExtension()!;
  expect(ext.getPoolNumber()).toBe('FN AS1234');
  expect(ext.getAgency()).toBe(AgencyProto.FNMA);
  expect(ext.getWac()?.getArbitraryPrecisionValue()).toBe('0.045');
  expect(ext.getWam()).toBe(358);
  expect(ext.getPassThroughRate()?.getArbitraryPrecisionValue()).toBe('0.04');
  expect(ext.getCurrentFactor()?.getArbitraryPrecisionValue()).toBe('0.95');
  expect(ext.getOriginalFaceValue()?.getArbitraryPrecisionValue()).toBe('250000000');
  expect(ext.getCurrentUpb()?.getArbitraryPrecisionValue()).toBe('237500000');
  expect(ext.getPsaSpeed()?.getArbitraryPrecisionValue()).toBe('150');
});

test('MortgageBackedSecurity wraps via Security.create factory dispatch', () => {
  const proto = MortgageBackedSecurity.fromPricerInputs(mbsInputs);
  const sec = Security.create(proto);
  expect(sec).toBeInstanceOf(MortgageBackedSecurity);
  // Inherits BondSecurity behaviour.
  expect(sec).toBeInstanceOf(BondSecurity);
  expect(sec.isBond()).toBe(true);
});

test('MortgageBackedSecurity typed accessors read back the expected values', () => {
  const proto = MortgageBackedSecurity.fromPricerInputs(mbsInputs);
  const sec = Security.create(proto) as MortgageBackedSecurity;

  // Bond-side checks (inherited from BondSecurity).
  expect(sec.getCouponRate().getArbitraryPrecisionValue()).toBe('0.04');
  expect(sec.getFaceValue().getArbitraryPrecisionValue()).toBe('250000000');
  expect(sec.getIssueDate().toDate().getFullYear()).toBe(2024);
  expect(sec.getMaturityDate().toDate().getFullYear()).toBe(2054);

  // MBS-specific accessor checks.
  expect(sec.getPoolNumber()).toBe('FN AS1234');
  expect(sec.getAgency()).toBe(AgencyProto.FNMA);
  expect(sec.getWac()?.toString()).toBe('0.045');
  expect(sec.getWam()).toBe(358);
  expect(sec.getPassThroughRate()?.toString()).toBe('0.04');
  expect(sec.getCurrentFactor()?.toString()).toBe('0.95');
  expect(sec.getOriginalFaceValue()?.toString()).toBe('250000000');
  expect(sec.getCurrentUpb()?.toString()).toBe('237500000');
  expect(sec.getPsaSpeed()?.toString()).toBe('150');
});
